text/0000-futures.md

```diff
@@ -154,31 +164,316 @@ can be dropped.
+        unsafe { (self.waker.vtable.wake)(self.waker.data) }
+    }
+
+    /// Converts the `LocalWaker` into `Waker`, which can be sent across
+    /// thread boundaries.
+    ///
+    /// This operation consumes the `LocalWaker`.
+    ///
+    /// This function can panic if the associated executor does not support
+    /// getting woken up from a different thread.
+    pub fn into_waker(self) -> Waker {
+        match self.try_into_waker() {
+            Some(waker) => waker,
+            None => panic!("Conversion from LocalWaker into Waker is not supported"),
+        }
+    }
+
+    /// Tries to convert the `LocalWaker` into a `Waker`, which can be sent
+    /// across thread boundaries.
+    ///
+    /// This operation consumes the `LocalWaker`.
+    ///
+    /// Returns None if the if the associated executor does not support
+    /// getting woken up from a different thread.
+    pub fn try_into_waker(self) -> Option<Waker> {
+        unsafe {
+            let maybe_raw_waker = (self.waker.vtable.into_waker)(self.waker.data);
+            // Avoid that the drop runs on self, which would e.g. decrease
+            // the refcount on it. The object has been consumed or converted
+            // by the `into_waker` function.
+            mem::forget(self);
+            match maybe_raw_waker {
+                Some(rw) => Some(Waker::new_unchecked(rw)),
+                None => None,
+            }
+        }
+    }
+
+    /// Returns whether or not this LocalWaker and other LocalWaker awaken the same task.
+    ///
+    /// This function works on a best-effort basis, and may return false even
+    /// when the LocalWakers would awaken the same task. However, if this function
+    /// returns true, it is guaranteed that the LocalWakers will awaken the same task.
+    ///
+    /// This function is primarily used for optimization purposes.
+    pub fn will_wake(&self, other: &LocalWaker) -> bool {
+        self.waker == other.waker
+    }
+
+    /// Creates a new `LocalWaker` from `RawWaker`.
+    ///
+    /// The method can not check whether `RawWaker` fulfills the required API
+    /// contract to make it usable for `LocalWaker` and is therefore unsafe.
+    pub unsafe fn new_unchecked(waker: RawWaker) -> LocalWaker {
+        LocalWaker {
+            waker: waker,
+        }
+    }
+}
+
+// The implementations of `Clone` and `Drop` follow what is shown in for `Waker`.
+```
+
+`Waker`s must fulfill the following requirements:
+- They must be cloneable
+- If all instances of a `Waker` have been dropped and their associated task had
+  been driven to completion, all resources which had been allocated for the task
+  must have been released.
+- It must be safe to call `wake()` on a `Waker` even if the associated task has
+  already been driven to completion.
+- `Waker::wake()` must wake up an executor even if it is called from an arbitry
+  thread.
+
+An executor which implements `RawWaker` must therefore make sure that all these
```

Comment:
A `RawWaker` as defined here is not a trait, so how can it be implemented?

Reply:
Yes the wording isn't ideal. Maybe something along `An executor which creates instances of RawWaker according to it's requirements`?

Reply:
Or just: "An executor that instantiates a `RawWaker` must therefore make sure..."

Reply:
I went with that